### DISCUSSION TITLE
Add support for System.Net.Http.Json `GetFromJsonAsync` methods

### DISCUSSION THREAD

**Discussion by eduherminio** [App-vNext/Polly#1411] [Ideas]
It would be awesome to be able to wrap with Polly this call:

```csharp
await _httpClient.GetFromJsonAsync<A>("relative_url", jsonSerializerOptions, cancellationToken);
```

Currently I haven't found a better way stop using `GetFromJsonAsync` extension method:

```csharp
var response = await _retryPolicy.ExecuteAsync(async (_) => await _httpClient.GetAsync("relative_url", cancellationToken), cancellationToken);

return JsonSerializer.Deserialize<A>(await response.Content.ReadAsStringAsync(cancellationToken), jsonSerializerOptions);
```

----

The simplest solution I could thing of is adding overrides to all `GetFromJsonAsync` methods [found here](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/HttpContentJsonExtensions.cs) with a `PolicyWrap` parameter?
Although we would be executing the serialization part inside of the `PolicyWrap.Execute(...)` method, not sure how dumb is that.,

Sample:
```csharp
public static partial class HttpClientJsonPollyExtensions
{
    internal const string SerializationUnreferencedCodeMessage = "JSON serialization and deserialization might require types that cannot be statically analyzed. Use the overload that takes a JsonTypeInfo or JsonSerializerContext, or make sure all of the required types are preserved.";
    internal const string SerializationDynamicCodeMessage = "JSON serialization and deserialization might require types that cannot be statically analyzed. Use the overload that takes a JsonTypeInfo or JsonSerializerContext.";

    [RequiresUnreferencedCode(SerializationUnreferencedCodeMessage)]
    [RequiresDynamicCode(SerializationDynamicCodeMessage)]
    public static Task<object?> GetFromJsonAsync(this HttpClient client, PolicyWrap policy, [StringSyntax(StringSyntaxAttribute.Uri)] string? requestUri, Type type, JsonSerializerOptions? options, CancellationToken cancellationToken = default)
    {
        return policy.Execute(() => client.GetFromJsonAsync(requestUri, type, options, cancellationToken));
    }

    // ...

}
```

**Comment by martincostello** [comment]
Have you looked into [using HttpClientFactory with Polly](https://learn.microsoft.com/dotnet/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests)? That plugs Polly v7 into _any_ HTTP request, not just ones using the method for getting JSON.

There will be further new APIs for integrating Polly v8 into HTTP pipelines coming from the .NET team via the https://github.com/dotnet/extensions repository as part of the .NET 8 release later this year.

**Reply by eduherminio** [reply]
That would make sense, but I was avoiding it for my use case since the general recommendation is to keep HttpClient instances generated by IHttpClientFactory short-lived (if I understood it correctly from https://learn.microsoft.com/en-us/dotnet/core/extensions/httpclient-factory#httpclient-lifetime-management) and I'm using this as a long-lived instance inside of a static class within a project that doesn't use DI.

**Reply by eduherminio** [reply]
IMHO as far as the recommendation is not to use IHttpClientFactory for long-lived clients (as in [HttpClient lifetime management](https://learn.microsoft.com/en-us/dotnet/fundamentals/networking/http/httpclient-guidelines#httpclient-lifetime-management)), it would be beneficial to have the proposed extensions.

**Reply by martincostello** [reply]
If you don't want to use HttpClientFactory, then you could still use the delegating handler that wraps calls with Polly from the Microsoft.Extensions.Http.Polly package, you'd just have to create the HTTP message handler chain yourself manually for your long-lived static client.

We wouldn't add extensions that would only benefit a subset of HTTP methods and/or content types when there is already functionality available to inject Polly into the HTTP pipeline to handle any possible request.

**Reply by eduherminio** [reply]
> If you don't want to use HttpClientFactory, then you could still use the delegating handler that wraps calls with Polly from the Microsoft.Extensions.Http.Polly package, you'd just have to create the HTTP message handler chain yourself manually for your long-lived static client.

Oh, thanks for pointing me out to that!
Being able to properly configure the HttpClient that way would make me totally agree with your position about the extensions I suggested.

Would something like this make sense to you regarding handler chaining, to combine the Policy one with the Sockets one needed to control the `PooledConnectionLifetime`?

```csharp
   AsyncRetryPolicy<HttpResponseMessage> retryPolicy;

    var client = new HttpClient(
        new PolicyHttpMessageHandler(retryPolicy)
        {
            InnerHandler = new SocketsHttpHandler() { PooledConnectionLifetime = TimeSpan.FromMinutes(15) }
        })
    {
        BaseAddress = new("...")
    };
```

**Reply by martincostello** [reply]
That seems correct to me.

**Reply by eduherminio** [reply]
Cool, thanks for your support Martin!
